cmd/thanos/sidecar.go

```diff
@@ -176,8 +189,10 @@ func runSidecar(
 				if err := m.UpdateLabels(iterCtx, logger); err != nil {
 					level.Warn(logger).Log("msg", "heartbeat failed", "err", err)
 					promUp.Set(0)
+					readinessProber.SetNotReady(err)
```

Comment:
Wonder if it's ready or rather healthy?

Reply:
From: https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/
Liveness/Healthiness:
```
Many applications running for long periods of time eventually transition to broken states, and cannot recover except by being restarted. Kubernetes provides liveness probes to detect and remedy such situations.
```
Readiness:
```
Sometimes, applications are temporarily unable to serve traffic. For example, an application might need to load large data or configuration files during startup, or depend on external services after startup. In such cases, you don’t want to kill the application, but you don’t want to send it requests either. Kubernetes provides readiness probes to detect and mitigate these situations. A pod with containers reporting that they are not ready does not receive traffic through Kubernetes Services.
```
I would add those definitions to probe `SetReady` and `SetHealth` to be clear, maybe?

I believe we can set healthiness instead of readiness here, but also not sure how crash will help here.... (: More over I would retry couple of times first (e.g 3) to ensure we are not flaky. Maybe after 3rd heartbeat fail we should set to not ready / not healthy?

Reply:
I'd say the sidecar can easily recover from this state when the Prometheus instance goes back up or becomes responsible again. As you said, I'm not sure either if the restart helps here somehow. But probably some issue where the restart would help could occur. 

Maybe set it to not ready right after first fail and unhealthy after for example as you suggested 3 failed health-checks in a row or bit more? Personally I'd wait at least 2m for example? That could be sufficient for even bigger instance of Prometheus to restart.

Adding the comments sounds as a good idea, will do.

Reply:
Changed to `NotReady` right after health-check on Prometheus fails and `NotHealthy` after 3 minutes of consequential fails of the health-check. I wanted to use the number of failed health-checks but the check interval is different on startup and later when the sidecar is already running so I stick to the time to be consistent.

Does this make sense to you? I'm bit afraid to set it unhealthy right after one fail.

Reply:
I would try to avoid this logic, as this is easily implementable in Kubernetes via:
```
timeoutSeconds
failureThreshold
initialDelaySeconds
```
Docs: https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/#configure-probes


Let's just make it simple.

So if there is no connection to Prometheus, we can't serve traffic, so it's definetely *NotReady*.

Now is it not healthy? IMO as we keep retrying for prometheus, we are healthy, but Prometheus isn't. So Prometheus healthchecks should restart it and make sure it's there and healthy, not sidecar's.

@bwplotka / @FUSAKLA  WDYT?

Reply:
Did not see that comment, sorry. will respond in more visible thread ;p